client/fingerprint/env_digitalocean.go

```diff
@@ -0,0 +1,173 @@
+package fingerprint
+
+import (
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"time"
+
+	cleanhttp "github.com/hashicorp/go-cleanhttp"
+	log "github.com/hashicorp/go-hclog"
+
+	"github.com/hashicorp/nomad/helper/useragent"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+const (
+	// DigitalOceanMetadataURL is where the DigitalOcean metadata api normally resides.
+	DigitalOceanMetadataURL = "http://169.254.169.254/metadata/v1/"
+
+	// DigitalOceanMetadataTimeout is the timeout used when contacting the DigitalOcean metadata
+	// services.
+	DigitalOceanMetadataTimeout = 2 * time.Second
+)
+
+type DigitalOceanMetadataPair struct {
+	path   string
+	unique bool
+}
+
+// EnvDigitalOceanFingerprint is used to fingerprint DigitalOcean metadata
+type EnvDigitalOceanFingerprint struct {
+	StaticFingerprinter
+	client      *http.Client
+	logger      log.Logger
+	metadataURL string
+}
+
+// NewEnvDigitalOceanFingerprint is used to create a fingerprint from DigitalOcean metadata
+func NewEnvDigitalOceanFingerprint(logger log.Logger) Fingerprint {
+	// Read the internal metadata URL from the environment, allowing test files to
+	// provide their own
+	metadataURL := os.Getenv("DO_ENV_URL")
+	if metadataURL == "" {
+		metadataURL = DigitalOceanMetadataURL
+	}
+
+	// assume 2 seconds is enough time for inside DigitalOcean network
+	client := &http.Client{
+		Timeout:   DigitalOceanMetadataTimeout,
+		Transport: cleanhttp.DefaultTransport(),
+	}
+
+	return &EnvDigitalOceanFingerprint{
+		client:      client,
+		logger:      logger.Named("env_digitalocean"),
+		metadataURL: metadataURL,
+	}
+}
+
+func (f *EnvDigitalOceanFingerprint) Get(attribute string, format string) (string, error) {
+	reqURL := f.metadataURL + attribute
+	parsedURL, err := url.Parse(reqURL)
+	if err != nil {
+		return "", err
+	}
+
+	req := &http.Request{
+		Method: "GET",
+		URL:    parsedURL,
+		Header: http.Header{
+			"User-Agent": []string{useragent.String()},
+		},
+	}
+
+	res, err := f.client.Do(req)
+	if err != nil {
+		f.logger.Debug("could not read value for attribute", "attribute", attribute, "error", err)
+		return "", err
+	} else if res.StatusCode != http.StatusOK {
+		f.logger.Debug("could not read value for attribute", "attribute", attribute, "resp_code", res.StatusCode)
+		return "", err
```

Comment:
This error could be `nil`; its value is unrelated to the status code

Reply:
I moved this branch lower in the function to include the resp body in the error message. Not 100% sure if this is safe (i.e. will digitalocean always return some body with an error response), but it would improve debugability if it does work